### ISSUE TITLE
Performance downgraded in the newer version compare to version before 1.x

### ISSUE THREAD

**Issue by zh1cheng** [gabime/spdlog#2294]
Today I tried to see whether to upgrade my project to use the latest version of spdlog. The current version I am using is the version before 1.x (forgot which version exactly), which still uses << to output logs.  The problem in the version before 1.x is that when using async log, there is always a string copy. So I am trying to see whether it still has this string copy in the latest version.

When trying to run the same code, I noticed the performance differences in the two versions. e.g. in the version before 1.x  about 8 messages per microseconds. however, in the latest version, it is about 3 messages per microsecond. The message/per microseconds was extracted from the timestamp in the log file. My PC is running on windows 10, CPU i7-8750H. 20G RAM DDR4.

Any idea whether anything is wrong with my way of measuring the performance? What could be the cause of the poor performance compared to the older version? Is there a way to avoid the string copy(coming from `async_msg(const details::log_msg& m)` function in file [details\async_log_helper.h](https://github.com/gabime/spdlog/blob/560df2878ad308b27873b3cc5e810635d69cfad6/include/spdlog/details/async_log_helper.h)) in the version before 1.x?

Below is my code for the two version:
Version before 1.x:
```
std::shared_ptr<spdlog::logger> g_logger;
#define LOG(level) if (g_logger != NULL) g_logger->level()

////////////////////////////////////////////////////////////
// in main function
spdlog::set_level(spdlog::level::trace);
spdlog::set_pattern("[%H:%M:%S.%f] %t %L %v");
g_logger = spdlog::daily_logger_mt("x", "./log/x", 0, 0, true);

for (int i =0 ; i < 10000; ++i)
	LOG(info)<<"this is an inform log "<<"  something here " <<i;

_getch();
g_logger->flush();
spdlog::drop_all();
```

Code for the latest version:
```
std::shared_ptr<spdlog::logger> g_logger;
#define LOG(level) if (g_logger != NULL) g_logger->level

////////////////////////////////////////////////////////////
// main function
spdlog::set_level(spdlog::level::trace);
spdlog::set_pattern("[%H:%M:%S.%f] %t %L %v");
g_logger = spdlog::daily_logger_mt<spdlog::async_factory>("x", "./log/x", 0, 0, true);

for (int i =0 ; i < 10000; ++i)
	LOG(info)("this is an inform log {0} {1:d} ",  " something here ", i);

_getch();
g_logger->flush();
spdlog::drop_all();
```

**Comment by zh1cheng** [comment]
Just an update on my finding.  I have changed the async_msg.txt into fmt::MemoryWriter type and changed the copy constructor into a move constructor/assignment when converting between log_msg and async_msg. There was no string copy anymore, and the performance improved from 8 messages per microsecond to about 11-12 messages per microsecond, using the measuring method I mentioned above.

But I am still curious why the older version's performance is better than the latest version?

**Comment by gabime** [comment]
I wouldn't call 3 million msg/sec “poor performance” .
PR to improve performance even more is always welcome.

**Comment by zh1cheng** [comment]
I agree. Maybe it is not proper to say the "Poor Performance".

But my question is why was the performance downgraded from 8 million msgs/sec to 3 million msgs/sec in the newer version?

**Comment by gabime** [comment]
Probably because the implementation of the queue was replaced. The lockless queue was replaced with mutex protected queue due to problematic cpu consumption when the queue was full or empty. See  #1973  and #207

**Comment by zh1cheng** [comment]
Checked the issue in #207， and I realized the code I am using in my project was something like below.  But I cannot remember the changes were done by me or it was already like this when downloading your code from Github.

My thoughts are the following:
1. We should try to avoid the queue in a full situation, so why not create a queue big enough? In my project, the queue size I configured to 65536. My log frequency is a few hundred msgs/ms, and it never encountered the queue full situation.

2. When the queue is empty, why cannot wait for another 50ms or 100ms to check again? I do not need the log to be flushed less than 1ms, especially in the async mode. So 100ms delay in flushing the message to file is acceptable to me.

``
inline void spdlog::details::async_log_helper::sleep_or_yield(const spdlog::log_clock::time_point& now, const spdlog::log_clock::time_point& last_op_time)
{
    using std::chrono::milliseconds;
    using namespace std::this_thread;

    auto time_since_op = now - last_op_time;

    // spin upto 1 ms
    //if (time_since_op <= milliseconds(1))
    //    return;

    // yield upto 10ms
    //if (time_since_op <= milliseconds(10))
    //    return yield();


    // sleep for half of duration since last op
    if (time_since_op <= milliseconds(100))
        return sleep_for(milliseconds(50));

    return sleep_for(milliseconds(100));
}
``
